File: src/app/utils/evRouteAlgorithm.js
'use client';

// Client-side version of the EV route algorithm
import { decode } from '@googlemaps/polyline-codec';

// Helper function to determine if code is running on the server
const isServer = typeof window === 'undefined';

// Import the server algorithm for re-export if needed
import { planEvRoute as serverPlanEvRoute } from './serverEvRouteAlgorithm';

/**
 * Makes a Google Maps API request using the appropriate method (direct or proxy)
 * @param {string} endpoint - API endpoint path (e.g., 'directions/json')
 * @param {Object} params - Query parameters for the API call
 * @returns {Promise<Object>} API response data
 */
async function makeGoogleMapsApiRequest(endpoint, params) {
  // Create URL search params
  const searchParams = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    searchParams.append(key, value);
  }
  
  try {
    let url;
    let response;
    
    if (isServer) {
      // Server-side: use direct fetch with API key
      url = `https://maps.googleapis.com/maps/api/${endpoint}?${searchParams.toString()}&key=${process.env.NEXT_PUBLIC_GOOGLE_MAPS_API_KEY}`;
      response = await fetch(url);
    } else {
      // Client-side: use our API proxy route
      searchParams.append('endpoint', endpoint);
      url = `/api/maps?${searchParams.toString()}`;
      response = await fetch(url);
    }
    
    return await response.json();
  } catch (error) {
    console.error(`Error making Google Maps API request to ${endpoint}:`, error);
    throw error;
  }
}

/**
 * Calculates distance between two coordinates using the Haversine formula
 * @param {[number, number]} coord1 - First coordinate [lat, lng]
 * @param {[number, number]} coord2 - Second coordinate [lat, lng]
 * @returns {number} Distance in kilometers
 */
function getDistance(coord1, coord2) {
  const [lat1, lon1] = coord1;
  const [lat2, lon2] = coord2;
  
  const R = 6371; // Earth's radius in km
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  
  const a = 
    Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * 
    Math.sin(dLon/2) * Math.sin(dLon/2);
  
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
  return R * c;
}

/**
 * Gets route path and time estimation using Google Maps Directions API
 * @param {[number, number]} start - Starting coordinates [lat, lng]
 * @param {[number, number]} end - Ending coordinates [lat, lng]
 * @returns {Promise<{path: Array<[number, number]>, duration: number}>} Path and duration
 */
async function getRoadPathAndTimeWithTraffic(start, end) {
  try {
    const [startLat, startLng] = start;
    const [endLat, endLng] = end;
    
    // Make API request through our helper function
    const data = await makeGoogleMapsApiRequest('directions/json', {
      origin: `${startLat},${startLng}`,
      destination: `${endLat},${endLng}`,
      mode: 'driving'
    });
    
    if (!data.routes || data.routes.length === 0) {
      return { path: [], duration: 0 };
    }
    
    let path = [];
    let totalDuration = 0;
    
    for (const leg of data.routes[0].legs) {
      totalDuration += leg.duration.value;
      
      for (const step of leg.steps) {
        const points = decode(step.polyline.points);
        path = [...path, ...points];
      }
    }
    
    return { 
      path: path.map(point => [point[0], point[1]]), 
      duration: totalDuration 
    };
  } catch (error) {
    console.error("Error getting directions:", error);
    return { path: [], duration: 0 };
  }
}

/**
 * Gets a point along a route at a specified distance
 * @param {Array<[number, number]>} routePoints - Array of route points
 * @param {number} distanceKm - Target distance in kilometers
 * @returns {[number, number]} Point along the route
 */
function getPointAlongRoute(routePoints, distanceKm) {
  let totalDistance = 0;
  
  for (let i = 1; i < routePoints.length; i++) {
    const segDistance = getDistance(routePoints[i - 1], routePoints[i]);
    totalDistance += segDistance;
    
    if (totalDistance >= distanceKm) {
      return routePoints[i];
    }
  }
  
  return routePoints[routePoints.length - 1];
}

/**
 * Searches for nearby EV charging stations
 * @param {[number, number]} location - Location coordinates [lat, lng]
 * @returns {Promise<Array>} Array of charging stations
 */
async function searchNearbyEvStations(location, initialRadius = 50000, maxRadius = 150000, step = 25000) {
  const [lat, lng] = location;
  let radius = initialRadius;
  
  while (radius <= maxRadius) {
    console.log(`Searching EV stations near [${lat}, ${lng}] within ${radius} meters...`);
    
    try {
      // Make API request through our helper function
      const data = await makeGoogleMapsApiRequest('place/nearbysearch/json', {
        location: `${lat},${lng}`,
        radius: radius.toString(),
        keyword: 'EV charging station',
        type: 'charging_station'
      });
      
      if (data.results && data.results.length > 0) {
        return data.results.map(place => ({
          name: place.name,
          location: [place.geometry.location.lat, place.geometry.location.lng],
          vicinity: place.vicinity || '',
          chargingSpeedKW: [30,45,60][Math.floor(Math.random() * 3)], // Random charging speed
          waitTimeMin: Math.floor(Math.random() * 26) + 5 // Random wait time between 5-30 min
        }));
      }
      
      radius += step;
    } catch (error) {
      console.error("Error finding charging stations:", error);
      return [];
    }
  }
  
  return [];
}

/**
 * Filters stations that are closer to destination than current location
 * @param {Array} stations - Array of charging stations
 * @param {[number, number]} current - Current coordinates [lat, lng]
 * @param {[number, number]} destination - Destination coordinates [lat, lng]
 * @returns {Array} Filtered stations
 */
function filterStationsTowardDestination(stations, current, destination) {
  return stations.filter(station => 
    getDistance(station.location, destination) < getDistance(current, destination)
  );
}

/**
 * Scores stations based on their proximity to destination
 * @param {Array} stations - Array of charging stations
 * @param {[number, number]} destination - Destination coordinates [lat, lng]
 * @returns {Array} Scored stations
 */
function scoreStations(stations, destination) {
  return stations
    .map(station => {
      const dist = getDistance(station.location, destination);
      return { ...station, score: 1 / (dist + 1e-5) };
    })
    .sort((a, b) => b.score - a.score);
}

/**
 * Plans an EV route
 * @param {[number, number]} start - Starting coordinates [lat, lng]
 * @param {[number, number]} destination - Destination coordinates [lat, lng]
 * @param {number} batteryPercentage - Initial battery percentage
 * @param {number} fullRangeKm - Maximum range in kilometers on full battery
 * @returns {Promise<Object>} Route information
 */
export async function planEvRoute(start, destination, batteryPercentage, fullRangeKm) {
  let current = start;
  const route = [start];
  let remainingBattery = batteryPercentage;
  const stationsVisited = [];
  
  let totalDriveTimeSec = 0;
  let totalChargingTimeMin = 0;
  let totalWaitTimeMin = 0;
  const batteryCapacityKWh = 60;
  
  const logs = [];
  
  function log(message) {
    console.log(message);
    logs.push(message);
  }
  
  while (getDistance(current, destination) > 0) {
    const maxRange = (remainingBattery / 100) * fullRangeKm * 0.9;
    
    log(`Current location: [${current[0].toFixed(4)}, ${current[1].toFixed(4)}]`);
    log(`Distance to destination: ${getDistance(current, destination).toFixed(2)} km`);
    log(`Battery: ${remainingBattery.toFixed(2)}%, Estimated range: ${maxRange.toFixed(2)} km (with 10% reserve)`);
    
    if (getDistance(current, destination) <= maxRange) {
      log("Destination is within range. Driving directly.");
      const { path, duration } = await getRoadPathAndTimeWithTraffic(current, destination);
      totalDriveTimeSec += duration;
      route.push(destination);
      break;
    }
    
    const { path: routePoints } = await getRoadPathAndTimeWithTraffic(current, destination);
    
    if (!routePoints || routePoints.length === 0) {
      log("No route found.");
      break;
    }
    
    const idealPoint = getPointAlongRoute(routePoints, maxRange);
    log(`Targeting a point ${maxRange.toFixed(2)} km away: [${idealPoint[0].toFixed(4)}, ${idealPoint[1].toFixed(4)}]`);
    
    const nearbyStations = await searchNearbyEvStations(idealPoint);
    
    if (!nearbyStations || nearbyStations.length === 0) {
      log("No EV stations found near this segment.");
      break;
    }
    
    const filtered = filterStationsTowardDestination(nearbyStations, current, destination);
    
    if (!filtered || filtered.length === 0) {
      log("No suitable charging stations on the way.");
      break;
    }
    
    log(`Found ${filtered.length} stations after filtering:`);
    filtered.forEach(station => {
      log(`- ${station.name} at [${station.location[0].toFixed(4)}, ${station.location[1].toFixed(4)}] (${station.vicinity})`);
    });
    
    const scoredStations = scoreStations(filtered, destination);
    const reachableStations = [];
    
    for (const station of scoredStations) {
      const distanceToStation = getDistance(current, station.location);
      const batteryRequired = (distanceToStation / fullRangeKm) * 100;
      
      if (batteryRequired <= remainingBattery) {
        reachableStations.push({
          ...station,
          batteryUsed: batteryRequired,
          distance: distanceToStation
        });
      }
    }
    
    if (!reachableStations || reachableStations.length === 0) {
      log("No reachable stations found.");
      break;
    }
    
    const best = reachableStations[0];
    log(`Travelling to: ${best.name} (${best.vicinity}) | Distance: ${best.distance.toFixed(2)} km | Battery used: ${best.batteryUsed.toFixed(2)}%`);
    
    const { path, duration } = await getRoadPathAndTimeWithTraffic(current, best.location);
    totalDriveTimeSec += duration;
    
    remainingBattery -= best.batteryUsed;
    
    const batteryToCharge = (100 - remainingBattery) / 100;
    const energyNeededKWh = batteryToCharge * batteryCapacityKWh;
    const chargingTimeHr = energyNeededKWh / best.chargingSpeedKW;
    const chargingTimeMin = chargingTimeHr * 60;
    
    totalChargingTimeMin += chargingTimeMin;
    totalWaitTimeMin += best.waitTimeMin;
    
    log(`Charging at ${best.name} | Charging time: ${chargingTimeMin.toFixed(1)} min | Wait: ${best.waitTimeMin} min`);
    
    current = best.location;
    remainingBattery = 100;
    stationsVisited.push(best);
    route.push(current);
  }
  
  // Final leg to destination if needed
  if (route[route.length - 1][0] !== destination[0] || route[route.length - 1][1] !== destination[1]) {
    const { path, duration } = await getRoadPathAndTimeWithTraffic(current, destination);
    totalDriveTimeSec += duration;
    route.push(destination);
  }
  
  const totalDriveTimeHr = totalDriveTimeSec / 3600;
  const totalChargingTimeHr = totalChargingTimeMin / 60;
  const totalWaitTimeHr = totalWaitTimeMin / 60;
  const totalTripTimeHr = totalDriveTimeHr + totalChargingTimeHr + totalWaitTimeHr;
  
  log("Trip Summary:");
  log(`Total Driving Time: ${totalDriveTimeHr.toFixed(2)} hrs`);
  log(`Total Charging Time: ${totalChargingTimeHr.toFixed(2)} hrs`);
  log(`Total Wait Time: ${totalWaitTimeHr.toFixed(2)} hrs`);
  log(`Total Trip Time: ${totalTripTimeHr.toFixed(2)} hrs`);
  
  return {
    route,
    chargingStops: stationsVisited,
    summary: {
      drivingTime: totalDriveTimeHr,
      chargingTime: totalChargingTimeHr,
      waitTime: totalWaitTimeHr,
      totalTime: totalTripTimeHr
    },
    logs
  };
}

// Re-export the server function for client use
export const planEvRoute = serverPlanEvRoute; 

File: public/ev_route_algorithm.py
import googlemaps
from geopy.distance import geodesic
import folium
import time
import random
import datetime
from googlemaps.convert import decode_polyline

gmaps = googlemaps.Client(key="AIzaSyA8FXr0bT_nN_oTae0S2f_cNPZj-zNQRlQ")

def get_road_path_and_time_with_traffic(start, end):
    directions = gmaps.directions(start, end, mode='driving')
    if not directions:
        return [], 0

    path = []
    total_duration = 0

    for leg in directions[0]['legs']:
        total_duration += leg['duration']['value']
        for step in leg['steps']:
            polyline = step['polyline']['points']
            path.extend(decode_polyline(polyline))

    return [(point['lat'], point['lng']) for point in path], total_duration

def get_point_along_route(route_points, distance_km):
    total_distance = 0
    for i in range(1, len(route_points)):
        seg_distance = geodesic(route_points[i - 1], route_points[i]).km
        total_distance += seg_distance
        if total_distance >= distance_km:
            return route_points[i]
    return route_points[-1]

def search_nearby_ev_stations(location, initial_radius=50000, max_radius=150000, step=25000):
    radius = initial_radius
    while radius <= max_radius:
        print(f"🔍 Searching EV stations near {location} within {radius} meters...")
        lat, lng = location
        places_result = gmaps.places_nearby(
            location=(lat, lng),
            radius=radius,
            keyword="EV charging station",
            type="charging_station"
        )
        stations = []
        for place in places_result.get("results", []):
            loc = place["geometry"]["location"]
            stations.append({
                "name": place["name"],
                "location": (loc["lat"], loc["lng"]),
                "vicinity": place.get("vicinity", ""),
                "charging_speed_kW": random.choice([40, 50, 150]),
                "wait_time_min": random.randint(5, 30)
            })
        if stations:
            return stations
        radius += step
    return []

def filter_stations_toward_destination(stations, current, destination):
    return [s for s in stations if geodesic(s["location"], destination).km < geodesic(current, destination).km]

def score_stations(stations, destination):
    for station in stations:
        dist = geodesic(station["location"], destination).km
        station["score"] = 1 / (dist + 1e-5)
    return sorted(stations, key=lambda x: x["score"], reverse=True)

def plan_ev_route(start, destination, battery_percentage, full_range_km):
    current = start
    route = [start]
    remaining_battery = battery_percentage
    stations_visited = []

    total_drive_time_sec = 0
    total_charging_time_min = 0
    total_wait_time_min = 0
    battery_capacity_kWh = 60

    while geodesic(current, destination).km > 20:
        max_range = (remaining_battery / 100) * full_range_km * 0.9

        print(f"\n📍 Current location: {current}")
        print(f"🎯 Distance to destination: {geodesic(current, destination).km:.2f} km")
        print(f"🔋 Battery: {remaining_battery:.2f}%, Estimated range: {max_range:.2f} km (with 10% reserve)")

        if geodesic(current, destination).km <= max_range:
            print(f"✅ Destination is within range. Driving directly.")
            path_segment, segment_time = get_road_path_and_time_with_traffic(current, destination)
            total_drive_time_sec += segment_time
            route.append(destination)
            break

        # Only get route for finding the ideal point — no need to add segment_time here
        route_points, _ = get_road_path_and_time_with_traffic(current, destination)
        if not route_points:
            print("❌ No route found.")
            break

        ideal_point = get_point_along_route(route_points, max_range)
        print(f"➡️ Targeting a point {max_range:.2f} km away: {ideal_point}")

        nearby_stations = search_nearby_ev_stations(ideal_point)
        if not nearby_stations:
            print("⚠️  No EV stations found near this segment.")
            break

        filtered = filter_stations_toward_destination(nearby_stations, current, destination)
        if not filtered:
            print("⚠️  No suitable charging stations on the way.")
            break

        print(f"🧭 Found {len(filtered)} stations after filtering:")
        for station in filtered:
            print(f"- {station['name']} at {station['location']} ({station['vicinity']})")

        scored_stations = score_stations(filtered, destination)
        reachable_stations = []
        for station in scored_stations:
            distance_to_station = geodesic(current, station["location"]).km
            battery_required = (distance_to_station / full_range_km) * 100
            if battery_required <= remaining_battery:
                station["battery_used"] = battery_required
                station["distance"] = distance_to_station
                reachable_stations.append(station)

        if not reachable_stations:
            print("❌ No reachable stations found.")
            break

        best = reachable_stations[0]
        print(f"🔋 Travelling to: {best['name']} ({best['vicinity']}) | Distance: {best['distance']:.2f} km | Battery used: {best['battery_used']:.2f}%")

        # Actually drive to the selected charging station now
        path_segment, segment_time = get_road_path_and_time_with_traffic(current, best["location"])
        total_drive_time_sec += segment_time

        remaining_battery -= best["battery_used"]

        battery_to_charge = (100 - remaining_battery) / 100
        energy_needed_kWh = battery_to_charge * battery_capacity_kWh
        charging_time_hr = energy_needed_kWh / best["charging_speed_kW"]
        charging_time_min = charging_time_hr * 60

        total_charging_time_min += charging_time_min
        total_wait_time_min += best["wait_time_min"]

        print(f"⚡ Charging at {best['name']} | Charging time: {charging_time_min:.1f} min | Wait: {best['wait_time_min']} min")

        current = best["location"]
        remaining_battery = 100
        stations_visited.append(best)
        route.append(current)

        time.sleep(1)

    # Final leg to destination if needed
    if route[-1] != destination:
        path_segment, segment_time = get_road_path_and_time_with_traffic(current, destination)
        total_drive_time_sec += segment_time
        route.append(destination)

    total_drive_time_hr = total_drive_time_sec / 3600
    total_charging_time_hr = total_charging_time_min / 60
    total_wait_time_hr = total_wait_time_min / 60
    total_trip_time_hr = total_drive_time_hr + total_charging_time_hr + total_wait_time_hr

    print("\n🧾 Trip Summary:")
    print(f"🚘 Total Driving Time: {total_drive_time_hr:.2f} hrs")
    print(f"🔋 Total Charging Time: {total_charging_time_hr:.2f} hrs")
    print(f"⏱️  Total Wait Time: {total_wait_time_hr:.2f} hrs")
    print(f"🕒 Total Trip Time: {total_trip_time_hr:.2f} hrs")

    return {
        "route": route,
        "charging_stops": stations_visited
    }


def generate_map(route_info):
    m = folium.Map(location=route_info["route"][0], zoom_start=7)

    folium.Marker(route_info["route"][0], tooltip="Start", icon=folium.Icon(color='green')).add_to(m)
    folium.Marker(route_info["route"][-1], tooltip="Destination", icon=folium.Icon(color='red')).add_to(m)

    for stop in route_info["charging_stops"]:
        folium.Marker(stop["location"], tooltip=f"{stop['name']}\nSpeed: {stop['charging_speed_kW']}kW", icon=folium.Icon(color='orange')).add_to(m)

    for i in range(len(route_info["route"]) - 1):
        segment_path, _ = get_road_path_and_time_with_traffic(route_info["route"][i], route_info["route"][i+1])
        folium.PolyLine(segment_path, color="blue", weight=4).add_to(m)

    m.save("ev_route_map.html")
    print("\n✅ Map saved as 'ev_route_map.html'. Open it in a browser.")

if __name__ == "__main__":
    start = (19.09076083476796, 72.86272114897756)
    destination = (28.557960485893343, 77.1000928375823)
    battery_percentage = 95
    full_range_km = 300

    result = plan_ev_route(start, destination, battery_percentage, full_range_km)

    print("\nPlanned Route:")
    for point in result["route"]:
        print(point)

    print("\nCharging Stops:")
    for stop in result["charging_stops"]:
        print(f"{stop['name']} at {stop['location']} ({stop['vicinity']})")

    generate_map(result)
